### PR TITLE
feat: Added AlignComponent layout component

### LINKED ISSUE #1421: Add onParentResize lifecycle method
Currently we have the `onGameResize()` lifecycle method which fires first before the component's `onLoad()`, and then every time the game canvas changes.

This FR is about creating an alternative lifecycle event `onParentResize()` that would be conceptually similar to `onGameResize`, but without the deficiencies of the latter. The `onGameResize` method will remain as-is, thus this FR will not introduce any breaking changes.

The logic of `onParentResize()` is as follows:
  - Whenever a component changes size, it will call `onParentResize()` for all its children.
  - If the child does NOT change its size, then `onParentResize()` is not propagated deeper into the tree.
  - This method is NOT invoked before `onLoad()`.
  - This method is invoked for the `FlameGame` class when the widget size changes.

### LINKED ISSUE #2302: Create LayoutComponent classes
# Proposal

`LayoutComponent` will be the base class for a family of utility classes which handle the layout of their children. This would be quite similar to Flutter layout widgets such as `Align`, `Column`, `Center`, `Container`, `Padding`, `SizedBox`, etc.

The purpose of these new components is to assist with constructing UIs within a game: dialogue boxes, HUD, informational screens, in-game UIs such as dialogue or trade windows, etc.

PR #1971 already attempts to add two such layout classes: `Row` and `Column` -- this proposal is to have more of them, all under a single umbrella of `LayoutComponent`.

For example, suppose you're making a game and you want to display current score in the top right corner. Then, you should be able to declare it like this:
```dart
camera.viewport.add(
  Padding(
    padding: EdgeInsets.all(20),
    children: [
      Align(
        anchor: Anchor.topRight,
        child: ScoreComponent(),
      ),
    ],
  ),
);
```
After this, the game would take care to properly position the `ScoreComponent` and move it whenever the game resizes.

## Implementation

`LayoutComponent` would rely on the `onParentResize` lifecycle method (#1421).

The tree structure can be either normal (the target is added as a child of the LayoutComponent), or flat (the target is added to the parent directly, and layout component is a sibling). I'm not sure which option is better here.

### PR DESCRIPTION
**Pull request by st-pasha**
# Description

- This PR adds first **layout component**: `AlignComponent`, an equivalent of [Align](https://api.flutter.dev/flutter/widgets/Align-class.html) widget in Flutter.
- `AlignComponent` sizes itself to its parent, and then keeps its child aligned to the specified anchor within its own bounding box.
- Also adding `onParentResize()` lifecycle method, which is similar to `onGameResize`, but fires whenever the parent of the current component changes its size for any reason. (`FlameGame` is assumed to have the size `canvasSize`, and will invoke `onParentResize` whenever the canvas size changes).
- Additional layout components are planned to be added in future PRs.

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [x] No, this PR is not a breaking change.

## Related Issues

Closes #2302
Closes #1421

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

File: examples/lib/stories/layout/align_component.dart [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    the shape of a diamond.
+    
+    The arrangement will remain intact if you change the window size.
+  ''';
+
+  @override
+  void onLoad() {
+    addAll([
+      AlignComponent(
+        child: CircleComponent(radius: 40)
```
Comment (spydon) [inline examples/lib/stories/layout/align_component.dart:19]: I think this would be clearer if the `children` argument was used instead of the cascading adds

File: packages/flame/lib/src/components/position_component.dart [hunk 2]
```diff
@@ -184,8 +185,16 @@ class PositionComponent extends Component
...
   /// recommended. Instead, in order to make the [PositionComponent] larger
   /// or smaller, change its [scale].
+  @override
   NotifyingVector2 get size => _size;
-  set size(Vector2 size) => _size.setFrom(size);
+
+  @override
+  set size(Vector2 size) {
+    _size.setFrom(size);
+    if (hasChildren) {
```
Comment (spydon) [inline packages/flame/lib/src/components/position_component.dart:194]: Does this check optimize anything? The foreach should also just do a check if it is empty right?

File: packages/flame/lib/src/camera/viewport.dart [hunk 3]
```diff
@@ -66,6 +66,9 @@ abstract class Viewport extends Component
       camera.viewfinder.onViewportResize();
     }
     onViewportResize();
+    if (hasChildren) {
```
Comment (spydon) [inline packages/flame/lib/src/camera/viewport.dart:69]: Does this check optimize anything? The foreach should also just do a check if it is empty right?

Comment (spydon) [inline packages/flame/lib/src/camera/viewport.dart:69]: Does this check optimize anything? The foreach should also just do a check if it is empty right?

File: packages/flame/lib/src/layout/align_component.dart [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+/// PlayerSprite().add(
+///   AlignComponent(
+///     child: HealthBar()..anchor = Anchor.bottomCenter,
+///     alignment: Anchor.topCenter,
+///     keepChildAnchor: true,
+///   ),
+/// );
+/// ```
+class AlignComponent extends PositionComponent {
+  /// Creates component that keeps its [child] positioned according to the
```
Comment (spydon) [inline packages/flame/lib/src/layout/align_component.dart:51]:
```suggestion
  /// Creates a component that keeps its [child] positioned according to the
```

**Review by spydon — COMMENTED** [review]
Very well written docs!

File: packages/flame/lib/src/components/position_component.dart (hunk 2 above)
Comment (st-pasha) [reply packages/flame/lib/src/components/position_component.dart:194]: `hasChildren` avoids instantiating a `ComponentSet` object in case it is null.

File: packages/flame/lib/src/layout/align_component.dart (hunk 4 above)
Comment (st-pasha) [reply packages/flame/lib/src/layout/align_component.dart]: done

File: examples/lib/stories/layout/align_component.dart (hunk 1 above)
Comment (st-pasha) [reply examples/lib/stories/layout/align_component.dart]: done

**Review by spydon — APPROVED** [review]

File: examples/lib/stories/layout/align_component.dart [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+              Vector2.all(25),
+              EffectController(
+                infinite: true,
+                duration: 0.75,
+                reverseDuration: 0.5,
+              ),
+            ),
+            AlignComponent(
+              alignment: Anchor.topCenter,
+              child: CircleComponent(radius: 10, anchor: Anchor.bottomCenter),
```
Comment (erickzanardo) [inline examples/lib/stories/layout/align_component.dart:32]: Nit: this and the other occurrences could benefit from a trailing comma

File: packages/flame/lib/src/camera/viewport.dart (hunk 3 above)
Comment (luanpotter) [reply packages/flame/lib/src/camera/viewport.dart:69]: +1

File: examples/lib/stories/components/game_in_game_example.dart [hunk 6]
```diff
@@ -36,9 +36,10 @@ class GameChangeTimer extends TimerComponent
...
     final child = gameRef.draggablesGame.square;
-    final newParent = child.parent == gameRef.draggablesGame
-        ? gameRef.composedGame.parentSquare
-        : gameRef.draggablesGame;
-    child.changeParent(newParent);
+    if (child.parent == gameRef.draggablesGame) {
+      child.parent = gameRef.composedGame.parentSquare;
+    } else {
+      child.parent = gameRef.draggablesGame;
+    }
```
Comment (luanpotter) [inline examples/lib/stories/components/game_in_game_example.dart:43]: is this related? we shouldn't use the method changeParent?

File: packages/flame/lib/src/components/position_component.dart (hunk 2 above)
Comment (luanpotter) [reply packages/flame/lib/src/components/position_component.dart:194]: could we do that check inside the `.children` getter instead?
that way we avoid duplicating this all over & everyone gets the performance boost (even end users)

File: packages/flame/lib/src/game/flame_game.dart [hunk 7]
```diff
@@ -182,4 +184,9 @@ class FlameGame extends ComponentTreeRoot with Game {
       }
     }
   }
+
+  @override
+  set size(Vector2 value) {
+    throw UnsupportedError('The size of FlameGame cannot be set explicitly');
+  }
```
Comment (luanpotter) [inline packages/flame/lib/src/game/flame_game.dart:191]: would it be beneficial to create another interface fot things that can only get the size?
like:
```dart
abstract class ReadableSizeProvider {
  Vector2 get size;
}
abstract class SizeProvider extends ReadableSizeProvider {
  // ...
}
```
then each class can implement specifically what they have, and things are a bit more type safe

File: packages/flame/lib/src/layout/align_component.dart [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+/// The component requires a single [child], which will be the target of this
+/// component's alignment. Of course, other children can be added to this
+/// component too, but only the initial [child] will be aligned.
+///
+/// The [alignment] parameter describes where the child should be placed within
+/// the current component. For example, if the [alignment] is `Anchor.center`,
+/// then the child will be centered.
+///
+/// Normally, this component's size will match the size of its parent. However,
+/// if you provide properties [widthFactor] or [heightFactor], then the size of
```
Comment (luanpotter) [inline packages/flame/lib/src/layout/align_component.dart:19]: should it be a factor: Vector2?

Comment (luanpotter) [reply packages/flame/lib/src/layout/align_component.dart:19]: or scaleFactor

File: packages/flame/lib/src/layout/align_component.dart [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+/// );
+/// ```
+///
+/// By default, the child's anchor is set equal to the [alignment] value. This
+/// achieves traditional alignment behavior: for example, the center of the
+/// child will be placed at the center of the current component, or bottom
+/// right corner of the child can be placed in the bottom right corner of the
+/// component. However, it is also possible to achieve more extravagant
+/// placement by giving the child a different anchor and setting
+/// [keepChildAnchor] to true. For example, if you set `alignment` to
```
Comment (luanpotter) [inline packages/flame/lib/src/layout/align_component.dart:38]: any reason you would _not_ want to keepChildAnchor?
why would you set an anchor to the child if you want to ignore it?

Comment (st-pasha) [reply packages/flame/lib/src/layout/align_component.dart:38]: Because when you receive a `child`, that child will have some non-null `anchor` value, and there is no way to know whether that anchor was set manually by the user or automatically by the framework.

File: packages/flame/lib/src/layout/align_component.dart (hunk 8 above)
Comment (st-pasha) [reply packages/flame/lib/src/layout/align_component.dart:19]: I tried to mimic here the api of the [Align](https://api.flutter.dev/flutter/widgets/Align-class.html) widget in Flutter.

Plus, having them as two separate parameters makes it easy to specify only one while leaving the other as `null`. For example, if you set `heightFactor: 1`, then it means a component that takes its height from the child, but the width from the parent.

File: examples/lib/stories/components/game_in_game_example.dart (hunk 6 above)
Comment (st-pasha) [reply examples/lib/stories/components/game_in_game_example.dart]: There was something weird going on with this snippet: my Dart decided that the common superclass between these RHSs is `Object`. The `changeParent` method can still be used, I reverted this change to almost the same as it was before.

File: packages/flame/lib/src/layout/align_component.dart [hunk 10]
```diff
@@ -0,0 +1,124 @@
+import 'package:flame/src/anchor.dart';
+import 'package:flame/src/components/position_component.dart';
+import 'package:flame/src/effects/provider_interfaces.dart';
+import 'package:flutter/widgets.dart';
+import 'package:vector_math/vector_math_64.dart';
+
+/// **AlignComponent** is a layout component that positions its child within
+/// itself using relative placement. It is similar to Flutter's [Align] widget.
+///
+/// The component requires a single [child], which will be the target of this
```
[the remark this answers is not in the document]
Comment (st-pasha) [reply packages/flame/lib/src/layout/align_component.dart:10]: It's should still be possible to add other children to this component, though they wouldn't be aligned. For example, you can add an `Effect` that would move the `AlignComponent` across the screen, or even change the `alignment` value over time.

File: packages/flame/lib/src/layout/align_component.dart [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+  AlignComponent({
+    required this.child,
+    required Anchor alignment,
+    this.widthFactor,
+    this.heightFactor,
+    this.keepChildAnchor = false,
+  }) {
+    this.alignment = alignment;
+    add(child);
+  }
```
[the remark this answers is not in the document]
Comment (st-pasha) [reply packages/flame/lib/src/layout/align_component.dart:66]: `this.alignment = alignment` ensures that the setter for the property is called.

File: examples/lib/stories/layout/align_component.dart (hunk 5 above)
Comment (st-pasha) [reply examples/lib/stories/layout/align_component.dart:32]: done

File: packages/flame/lib/src/game/flame_game.dart (hunk 7 above)
Comment (st-pasha) [reply packages/flame/lib/src/game/flame_game.dart:191]: Good idea

**Review by spydon — APPROVED** [review]

File: examples/lib/stories/components/game_in_game_example.dart (hunk 6 above)
Comment (luanpotter) [reply examples/lib/stories/components/game_in_game_example.dart]: sg, I was just curious if it was being removed or something

File: packages/flame/lib/src/layout/align_component.dart (hunk 9 above)
Comment (luanpotter) [reply packages/flame/lib/src/layout/align_component.dart:38]: ah! got it :)

File: packages/flame/lib/src/components/position_component.dart (hunk 2 above)
Comment (luanpotter) [reply packages/flame/lib/src/components/position_component.dart:194]: (can be a separate PR of course)

**Review by luanpotter — APPROVED** [review]
LGTM, nice addition!

File: packages/flame/lib/src/components/position_component.dart (hunk 2 above)
Comment (st-pasha) [reply packages/flame/lib/src/components/position_component.dart:194]: The `.children` property returns a `ComponentSet` right now, which means it must create a component set in case it is currently missing. We could change it to return a `ComponentSet?` instead, but that would be a major breaking change I'm afraid -- so perhaps we could add that to our 2.0 todo list.

Comment (spydon) [reply packages/flame/lib/src/components/position_component.dart:194]: Haha I think 3 people commented about this on this PR, maybe it would be better as nullable indeed, let's discuss that for V2 at least. :)

Comment (st-pasha) [reply packages/flame/lib/src/components/position_component.dart:194]: I added a note about this in #1938

**Comment by rivella50** [comment]
Are there plans to implement additional layout components (mainly for row and column support)?
As you probably remember i did some [work](https://github.com/flame-engine/flame/pull/1971) on this topic last year and perhaps it could be used here.

**Comment by spydon** [comment]
@rivella50 There is definitely some stuff in there that could be re-used, the main difference now is that the Row and Column components should be based on the `AlignComponent`. If you want to have a go at this I can assign you to the issue? :)

**Comment by rivella50** [comment]
@spydon I pass this time, thanks. When i worked on it the last time it was quite frustrating and seemed to turn into a neverending story since all the time another guy appeared on the stage and suggested new things and changes.
This just took me too much time and there was no end foreseeable.

**Comment by spydon** [comment]
@rivella50 makes sense, must have been frustrating!
